Fix double instantiation of widgets

shobhitagarwal1612 opened the pull request:
Closes #2537 

<!-- 
Thank you for contributing to ODK Collect!

Before sending this PR, please read
https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md
-->

#### What has been done to verify that this works as intended?
Added logs at the widget creation point to verify that only one instance is being generated when opening the form for the first time or returning from FormHierarchy Screen.

Test Cases verified:
 - Opening a fresh form
 - Opening a saved form
 - Opening a form > Goto > Choosing a widget
 - Opening a form > General Settings > Back
 - Opening a form > General Settings > Navigation changed to button > Back > Button visible

#### Why is this the best possible solution? Were any other approaches considered?
This is the only approach I considered since it seemed logical to me

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Does not change any behavior for the user. But it might speed up the response time after coming back from a hierarchy screen.

#### Do we need any specific form for testing your changes? If so, please attach one.
I tested with the All-Widgets form provided on the Aggregate server

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

lognaturel commented:
The `refreshCurrentView` call in `onResume` was added in 72a68710425283ecdf7612bab9a5da10177acfc4 (and later approved by Mitch), apparently to fix an issue related to the hierarchy view. It looks like the issue was that the view wasn't refreshed at all? 

Before your change, that statement is reached any time `formLoaderTask` is `null` and `formController` is not `null`. Can you please include your reasoning for determining it's safe to remove? (that is, what are the cases in which those conditions would occur and are they all handled elsewhere now?)

Am I understanding correctly that the double widget creation doesn't actually happen in master currently, it's when `startActivity` is replaced with `startActivityForResult` that it starts happening because `onActivityResult` also refreshes the view?

lognaturel commented:
To be explicit, I'm trying to figure out whether this is actually something that possibly affects users today or not and also trying to evaluate the level of risk. We have a code freeze in about 36 hours so those things will help figure out whether this should be included in the next release or not.

We'll then need to have a broader strategic discussion about #2537 and when it makes sense to fit in.

shobhitagarwal1612 commented:
> Am I understanding correctly that the double widget creation doesn't actually happen in master currently

It does happen in master currently

The reason for the hack was that, if the hierarchy view is opened directly (as in case of edit saved form), the hierarchy view activity was being started with `startActivity` instead of `startActivityForResult`. So as a result, after coming back to FormEntryActivity, `onActivityResult` wasn't getting called, only onResume was

shobhitagarwal1612 commented:
Another option to solve the duplication is to keep the hack and remove the refreshCurrentView call from onActivityResult.

shobhitagarwal1612 commented:
I do believe that this has been affecting users as all of the calls are synchronous and this adds up to the time taken to finally load the widget and interact with it

shobhitagarwal1612 commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -338,18 +338,18 @@ public void granted() {
                      */
                     if (onResumeWasCalledWithoutPermissions) {
                         onResume();
+                        refreshCurrentView();
```
This is only required when the form gets loaded for the first time. From then onwards, only onActivityResult is concerned to generate new widgets

lognaturel replied:
Won't `onResumeWasCalledWithoutPermissions` be false here always when the activity is first loaded because `onCreate` gets called before `onResume`? What do you mean by "for the first time"?

lognaturel replied:
Perhaps this code could benefit from some restructuring. I don't think it's generally a good idea to call lifecycle methods directly and this means `onResume` ends up being called twice. It would be better for the work that needs to be done in this case to be pulled out into a method. That might make the intent of the code clearer -- I'm no longer sure what this case is for exactly.

shobhitagarwal1612 replied:
I totally agree that this class needs refactoring. Infact I have done a lot of refactoring in the PR https://github.com/opendatakit/collect/pull/2535. This branch was only to apply a patch to the problem with least number of changes. But if you think it's a better idea to fix this problem along with code refactor, then I'd be more than happy to do that too

shobhitagarwal1612 commented:
> apparently to fix an issue related to the hierarchy view. It looks like the issue was that the view wasn't refreshed at all?

This issue was fixed by starting the hierarchy view for result, which then later gets intercepted by onActivityResult (refreshCurrentView generates the widget and refreshes the view)

lognaturel commented:
> The reason for the hack was that, if the hierarchy view is opened directly (as in case of edit saved form), the hierarchy view activity was being started with `startActivity` instead of `startActivityForResult`. So as a result, after coming back to `FormEntryActivity`, `onActivityResult` wasn't getting called, only `onResume` was

How does your fix address that case?

> Another option to solve the duplication is to keep the hack and remove the refreshCurrentView call from onActivityResult.

We need to have a clear explanation for why a particular approach is selected and what its implications are so any approach with those properties is ok.

lognaturel commented:
One quick idea to help document what the change improves and that it works in all cases is to put together a table that shows what happens on master and what happens with the fix when a user launches the hierarchy view from a form and then returns to form filling, launches the hierarchy view directly (e.g. from Edit Saved Form or opening a form with an automatic savepoint after a crash), etc. I use https://ozh.github.io/ascii-tables/ to make tables easily (shoutout to @danbjoseph for that tip). You've probably traced all those cases so hopefully it's quick for you.

shobhitagarwal1612 commented:
|                                            Action                                            | master | issue-2537 |
|----------------------------------------------------------------------------------------------|--------|------------|
| Open a new form > Loading of the first widget                                                | once   | once       |
| Open a last saved form > HierarchyActivity opens by default > Select any widget        | once   | once       |
| Open any form > Click Goto (Hierarchy Menu opens) > Select any widget (except above cases)  | twice  | once       |

shobhitagarwal1612 commented:
> How does your fix address that case?

Now that I am starting `EditHierarchyActivity` using `startActivityForResult`, so when returning from that activity, the below switch case gets handled https://github.com/opendatakit/collect/blob/d2b4dd38f6b305154169ab5f273efc7ed5eede5d/collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java#L866

And at the end of the method, the `refreshCurrentView` is called which refreshes the view (fix for the hack)

lognaturel commented:
For the table, the goal for me would be to summarize all the ways to get to the code paths that were changed. So what happens if storage permissions aren't granted on activity launch before and after the change? And the details in the cells would be the calls that happen and how that changes. I will try to get to tracing it so I can answer those questions.

> Now that I am starting `EditHierarchyActivity` using `startActivityForResult`, so when returning from that activity, the below switch case gets handled

The key here that I think is not obvious for someone who hasn't been in this code recently is that even when the hierarchy view is triggered from the main menu, it ends up going through the `FormEntryActivity`. The `start` extra makes sure the hierarchy view is shown and that's the block this code change is in. So even though this change is in `FormEntryActivity`, it's narrowly targeting the second case in your table.

> Infact I have done a lot of refactoring in the PR #2535. This branch was only to apply a patch to the problem with least number of changes. But if you think it's a better idea to fix this problem along with code refactor

If there are structural changes that group well with this change and make it clearer what's going on, I think it's a good idea to do them here. #2535 is massive, risky, and very difficult to review. If you can pull out some parts of it into smaller PRs, that will de-risk it. In this case, cleaning up the case where `onResume` was called without permissions seems like it might help.

I know this is tedious for a seemingly small change and I apologize! We can't afford to break things here. As a reviewer, ideally I have to fully understand the change or at the very least, I have to be certain that you understand the implications 100%.

shobhitagarwal1612 commented:
@lognaturel @grzesiek2010 It is ready for review

shobhitagarwal1612 commented:
> If there are structural changes that group well with this change and make it clearer what's going on, I think it's a good idea to do them here. #2535 is massive, risky, and very difficult to review. If you can pull out some parts of it into smaller PRs, that will de-risk it. In this case, cleaning up the case where onResume was called without permissions seems like it might help.

@lognaturel I have opened three PRs #2567 #2568 #2569 which pulls code from #2535 to make the review of #2535 easier and less risky

lognaturel commented:
I like the new commit messages!

What's the process you used to determine you don't need `onResume` at all? 

I believe in this case it's important to have `onResume` to handle the case where settings are changed from the overflow menu. For example, try going to General Settings > User interface > Navigation, changing the navigation type and then returning to filling the form. I believe you'll see this works well on `master` but that the navigation type won't have changed with this PR. I believe the `allowMovingBackwards` permission read should also be in `onResume`. It would be good to have a comment about this possibility of changing settings mid-form.

If you are confident the rest of the form loading work can be done from `onCreate` rather than `onResume`, it would be great to have an explanation for that.

shobhitagarwal1612 commented:
> I believe in this case it's important to have onResume to handle the case where settings are changed from the overflow menu. For example, try going to General Settings > User interface > Navigation, changing the navigation type and then returning to filling the form. I believe you'll see this works well on master but that the navigation type won't have changed with this PR. 

Fixed

> I believe the allowMovingBackwards permission read should also be in onResume. 

This preference can only be changed from Admin Preferences. So, it cannot be changed while filling in the form.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -337,19 +336,9 @@ public void granted() {
-                    if (onResumeWasCalledWithoutPermissions) {
-                        onResume();
-                    }
+                    setup();
```
Why is this called here? In normal execution, won't it just be called again in `onResume`?

shobhitagarwal1612 replied:
I haven't changed the original logic here. Let me check what is the impact of only calling it from `onResume`

shobhitagarwal1612 commented:
@lognaturel I've updated the PR with a better solution.

Refreshing the widget in `onResume()` should not be removed. This is because if we minimize the app while filling out a form and then reopen it, then the widget should refresh. Also, since `onResume()` gets called after `onActivtiyResult()` so removing the `refreshCurrentView()` from `onActivityResult()` seemed better to me.

But in case of saving media in a background thread (image, audio, video, file), we need to refresh the widget only after the media is loaded. So, only in this case, the `refreshCurrentView` shouldn't be called from `onResume()` but after the media is saved

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -867,9 +871,9 @@ public void run() {
         }
-        refreshCurrentView();
+
+        // now onResume() gets called and calls refreshCurrentView() to reload current widget
```
For the sake of completeness, I think you can add something like "If media is being loaded on a background thread, the refresh will be postponed until the media fully loads."

shobhitagarwal1612 replied:
@lognaturel Done

lognaturel commented:
Aha! Thanks for sticking with it -- this approach is looking a lot better to me. 🎉 

In addition to the cases you listed in the PR description, other cases to verify are loading large images and video from the chooser both locally and from Google Drive.  If you have a moment to update the PR description to match your latest understanding, that would be great!

shobhitagarwal1612 commented:
Yeah sure

shobhitagarwal1612 commented:
@mmarciniak90 To test this branch, I will suggest you add a log message to the class `WidgetFactory` on line 58. Then every time a new widget gets created, a log would be registered in the logcat.

mmarciniak90 commented:
@shobhitagarwal1612 thanks for the advice, it was very helpful!
Cases with 'Go To Start' and jumping to the selected widget are fixed.

I noticed that this behavior is also visible when I change device orientation and it was not fixed. 
@shobhitagarwal1612 should this case be also fixed as a part of this issue?

The second thing is Android option 'Do not keep activities'. When I have it turned on and I used the app without your fix new widget was created 3 times. 
On app with your fix, the new widget was created 2 times. 
Is it expected?

shobhitagarwal1612 committed:
Remove calling 'refreshCurrentView()' after receiving intent result

Do not call 'refreshCurrentView()' in onResume() if we are already saving image, video, audio or file widget. This is because we need to refresh the widget view after the media is loaded.

shobhitagarwal1612 commented:
> I noticed that this behavior is also visible when I change device orientation and it was not fixed.
@shobhitagarwal1612 should this case be also fixed as a part of this issue?
The second thing is Android option 'Do not keep activities'. When I have it turned on and I used the app without your fix new widget was created 3 times. On app with your fix, the new widget was created 2 times.
Is it expected?

@mmarciniak90 No, it's not an expected behavior, I missed this use-case. Thanks for catching this! 
I've fixed this in the latest commit

shobhitagarwal1612 committed:
Do not refresh the widget view when attempting to load the form

grzesiek2010 commented:
> I've fixed this in the latest commit

your last commit causes that e.g. it's not possible to take a photo.

shobhitagarwal1612 commented:
> your last commit causes that e.g. it's not possible to take a photo.

@grzesiek2010 Can you please explain this statement? Does this mean that the behavior was a result of the changes or does it mean that image capture isn't working anymore?

I have tried choosing/capturing image as well as rotating the orientation of the device and it worked fine. Am I missing something here?

lognaturel commented:
I'm also not able to reproduce, @grzesiek2010, and the logic seems sound.

grzesiek2010 commented:
@shobhitagarwal1612 @lognaturel 
I'm on this branch https://streamable.com/7cmhu
Everything is fine if I remove the last commit.

lognaturel commented:
@grzesiek2010 it would be helpful to know more about device, Android version and what exactly you're doing. Are you changing orientation? I'm still not able to reproduce.

shobhitagarwal1612 commented:
@grzesiek2010 I am unable to open this link https://streamable.com/7cmhu

grzesiek2010 commented:
I don't change device orientation I just try to take a photo in the simplest way it's possible.
Android 5
Samsung Galaxy S4
strange... 

@shobhitagarwal1612 what about this http://tinypic.com/player.php?v=mv2jgy%3E&s=9#.W873IHV94UE

shobhitagarwal1612 commented:
@grzesiek2010 I tested this branch on Motorola G5 Plus (Android 8.1) which does the extra rotation (like you mentioned) when returning back from DrawActivity not when capturing photos, but I did not notice any problem as mentioned in your comment.

I don't have access to any Samsung device. Is it possible for you to post the logs from your Samsung Galaxy S4 device on this thread?

grzesiek2010 commented:
It's difficult to understand the problem reading logs so I debugged the project.
The problem is that: [getCurrentViewIfODKView().setBinaryData(nf);](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java#L753) is not called because `getCurrentViewIfODKView()` returns null.

In my case it works in that way:
1. After accepting a photo I captured I'm moving back to ODK Collect
2. The activity is recreated so `currentView` field is null.
3. We try to save a captured photo but as I said above `getCurrentViewIfODKView()` returns null because `currentView` field is null.

If I keep the line you deleted in your last commit then refreshCurrentView() is called between steps 1 and 2 and `currentView` field is NOT null what allows us to save a captured photo.

IMPORTANT!
The problem is visible only if I use my camera app so `ImageWidget` and `AnnotateWidget` are affected. Widgets like `DrawWidget`, `SignatureWidget` even `SelfieWidget` (which doesn't use that camera) work well.

I guess it's because the camera app I use performs that additional recreation but it should be possible do such an orientation change manually as well and reproduce the issue using other apps but it might be very difficult. 

I tried to attach the camera name or find it on Play Store but no avail it's no name camera but I think not rare since it was a default app on my Samsung Galaxy S4.

shobhitagarwal1612 committed:
Ensure that getCurrentViewIfODKView() is not null before trying to
update binary data

Calling refreshCurrentView() creates the widget which gets assigned to the currentView variable. Since this ensures that the currentView is not null, so we can remove the null check from below it as well.

shobhitagarwal1612 commented:
> It's difficult to understand the problem reading logs so I debugged the project.
The problem is that: getCurrentViewIfODKView().setBinaryData(nf); is not called because getCurrentViewIfODKView() returns null.

Thanks @grzesiek2010! Based on your observation, I've added a check to ensure that the currentView isn't null before trying to set the binary data. I have checked that this does not have any negative impact for the normal cases (multiple creation of widget). Can you please try out the latest pushed commit on your device to check if this solves the problem?

grzesiek2010 commented:
No, it's not a solution.

I said: 
> The activity is recreated so currentView field is null

you just removed checking if it's null and tries to use that filed so it's going to cause NPEs

Also, I don't think it's a bug. You can't reproduce the issue because in most cases when you capture a photo and go back to the app `FormEntryActivity` is not recreated (`currentView` is not null as well then).
But if the activity is recreated then you are trying to call `setBinaryData()` on null values (not created again yet).

Maybe my camera app is not really common and other apps don't cause that recreation but I'm sure it's not only related to that camera app.

For example, what's important if you have a device with low memory it's possible that when you start a camera app `FormEntryActivity` is killed and then the same problem occurs.
I was able to reproduce that using one of my devices and enabling `Don't keep activities` option in `Developer settings` to simulate the mentioned case. 
Using that device I was able to take a photo when `Don't keep activities` option was disabled and the mentioned above problem (the same as on Samsung Galaxy S4) was visible when the option was enabled.

In my opinion is that we try to call `setBinaryData()` too early when the activity is not created yet again.

grzesiek2010 commented:
So short sum up is:
You can't reproduce the issue because in most cases `FormEntryActivity` is not killed after opening a camera app.
But if you use a device with low memory or you use the `Don't keep activities` option to simulate the same you should be able to reproduce the issue.

Maybe don't be focused on the camera app I used because maybe my device has low memory and that's why the issue occurs as I described above. Maybe it's not because of the app I can't be sure.
The case with `Don't keep activities` is easier to reproduce.

grzesiek2010 commented:
So the general problem is that you have to be sure that when `onActivityResult()` is called our layout is created and you can call `setBinaryData()`. That's why we used that line you removed in your second commit. Maybe it wasn't an elegant and meaningful way but at least it as always called before `onActivityResult()`.

lognaturel commented:
I think @grzesiek2010 is getting to a really good way to go about thinking through these kinds of changes -- what are the contexts in which this code needs to work and what's the best way to achieve that result. 

In this case, when does the view actually need to get rebuilt (because excessive refreshes are what is causing the double instantiation)? @grzesiek2010 has described that it might need to happen after focus comes back to the app and `FormEntryActivity` has been destroyed which is quite a subtle case but definitely an important one to handle. @shobhitagarwal1612 listed some other cases [here](https://github.com/opendatakit/collect/pull/2538#issuecomment-425165345). What other cases are there? Can that help figure out where the refresh should actually happen?

shobhitagarwal1612 commented:
@grzesiek2010 
Thanks for the advice. I was able to reproduce the bug on Motorola G5 Plus (Android 8.1) with `Do not keep activities` enabled.
I can confirm that 77ec599313236dd6a24b163c5f240add07abfad4 has the image capture bug and a68f3b660501a20acf06f8a2208959dd20434964 (or 9fe36ad93c3b6d8e27e8c8f45e59f007db96f43c the latest commit) solves that.

> No, it's not a solution.

Can you please confirm if you have tried out this branch out on your device and observed that the fix isn't working?

shobhitagarwal1612 committed:
Call refreshCurrentView() only when the variable is required

In the last commit, I added the null check for getCurrentViewIfODKView() in onActivityResult() but that initializes the widget even when it is not needed (eg. Moving back from HierarchyActivity). So, this creates another instance of the widget which we are trying to solve.

The best approach I could come up with was to instantiate the widget if "currentView" is null. This was required when "Do not keep activities" was enabled on Motorola G5 Plus/Samsung Galaxy S4 and we try to click a photo using the default camera app

grzesiek2010 commented:
I missed this https://github.com/opendatakit/collect/commit/9fe36ad93c3b6d8e27e8c8f45e59f007db96f43c#diff-03ff45091550285dd88b56927010c457R2662

I'm not sure if I can call it a fix. It's rather a workaround. The general approach is rather not good. Do you understand what is the root cause? It's a race condition problem that we try to setBinaryData() in onActivityResult when currentView is not ready. I would prefer to analyze it again deeply and try to fix the root cause.

shobhitagarwal1612 commented:
@grzesiek2010 I tried debugging the issue without using the last added workaround.

Hierarchy View -> Image Widget -> Take Picture -> Use camera app to capture a photo -> Return to app

I added log to some of the important methods and filtered the logcat based on the unique tag.

Logs collected:
-----------------------------------------
Selected image appears with `Do not keep activities` disabled:
```
D/FormEntryActivity: onResume() called
D/FormEntryActivity: refreshCurrentView() called
D/FormEntryActivity: releaseOdkView() called
D/WidgetFactory: createWidgetFromPrompt() called with: fep = [org.javarosa.form.api.FormEntryPrompt@ae52dcb], context = [org.odk.collect.android.activities.FormEntryActivity@1eba2e1], readOnlyOverride = [false]
I/FormEntryActivity: Created view for group Image widgets Image widget
D/FormEntryActivity: showView() called with: next = [org.odk.collect.android.views.ODKView{9243faa V.E...... ......I. 0,0-0,0}], from = [FADE]
D/FormEntryActivity: onPause() called
D/FormEntryActivity: onActivityResult() called with: requestCode = [1], resultCode = [-1], intent = [Intent {  }]
I/FormEntryActivity: Renamed /storage/emulated/0/odk/.cache/tmp.jpg to /storage/emulated/0/odk/instances/All widgets_2018-11-08_21-29-05/1541693414511.jpg
D/FormEntryActivity: onResume() called
D/FormEntryActivity: refreshCurrentView() called
D/FormEntryActivity: releaseOdkView() called
D/WidgetFactory: createWidgetFromPrompt() called with: fep = [org.javarosa.form.api.FormEntryPrompt@3508f44], context = [org.odk.collect.android.activities.FormEntryActivity@1eba2e1], readOnlyOverride = [false]
I/FormEntryActivity: Created view for group Image widgets Image widget
D/FormEntryActivity: showView() called with: next = [org.odk.collect.android.views.ODKView{8585b0c V.E...... ......I. 0,0-0,0}], from = [FADE]
```
------------------------------------------------------------------------
Selected image does not appear with `Do not keep activities` enabled
```
D/FormEntryActivity: onCreate() called with: savedInstanceState = [Bundle[{saveName=null, com.google.firebase.analytics.screen_service=Bundle[mParcelledData.dataSize=124], android:viewHierarchyState=Bundle[mParcelledData.dataSize=2612], instancepath=/storage/emulated/0/odk/instances/All widgets_2018-11-08_21-29-05/All widgets_2018-11-08_21-29-05.xml, error=null, xpath=question./all-widgets/image_widgets[1]/image_widget[1], android:savedDialogs=Bundle[mParcelledData.dataSize=2204], formpath=/storage/emulated/0/odk/forms/All widgets.xml, readPhoneStatePermissionRequestNeeded=false, android:lastAutofillId=1073741823, android:fragments=android.app.FragmentManagerState@855d646, autosaved=false, newform=false}]]
I/FormEntryActivity: startingXPath is: question./all-widgets/image_widgets[1]/image_widget[1]
D/FormEntryActivity: onActivityResult() called with: requestCode = [6], resultCode = [-1], intent = [null]
D/FormEntryActivity: onResume() called
D/FormEntryActivity: refreshCurrentView() called
D/FormEntryActivity: releaseOdkView() called
D/WidgetFactory: createWidgetFromPrompt() called with: fep = [org.javarosa.form.api.FormEntryPrompt@cb133c4], context = [org.odk.collect.android.activities.FormEntryActivity@3f9c007], readOnlyOverride = [false]
I/FormEntryActivity: Created view for group Image widgets Image widget
D/FormEntryActivity: showView() called with: next = [org.odk.collect.android.views.ODKView{fa8ceea V.E...... ......I. 0,0-0,0}], from = [FADE]
D/FormEntryActivity: onPause() called
D/FormEntryActivity: releaseOdkView() called
D/FormEntryActivity: onCreate() called with: savedInstanceState = [Bundle[{saveName=null, com.google.firebase.analytics.screen_service=Bundle[mParcelledData.dataSize=124], android:viewHierarchyState=Bundle[mParcelledData.dataSize=2536], instancepath=/storage/emulated/0/odk/instances/All widgets_2018-11-08_21-29-05/All widgets_2018-11-08_21-29-05.xml, xpathwaiting=question./all-widgets/image_widgets[1]/image_widget[1], error=null, xpath=question./all-widgets/image_widgets[1]/image_widget[1], android:savedDialogs=Bundle[mParcelledData.dataSize=2204], formpath=/storage/emulated/0/odk/forms/All widgets.xml, readPhoneStatePermissionRequestNeeded=false, android:lastAutofillId=1073741823, android:fragments=android.app.FragmentManagerState@998a2, autosaved=false, newform=false}]]
I/FormEntryActivity: startingXPath is: question./all-widgets/image_widgets[1]/image_widget[1]
I/FormEntryActivity: waitingXPath is: question./all-widgets/image_widgets[1]/image_widget[1]
D/FormEntryActivity: onActivityResult() called with: requestCode = [1], resultCode = [-1], intent = [Intent {  }]
E/FormEntryActivity: currentView returned null.
I/FormEntryActivity: Renamed /storage/emulated/0/odk/.cache/tmp.jpg to /storage/emulated/0/odk/instances/All widgets_2018-11-08_21-29-05/1541693446264.jpg
D/FormEntryActivity: onResume() called
D/FormEntryActivity: refreshCurrentView() called
D/FormEntryActivity: releaseOdkView() called
D/WidgetFactory: createWidgetFromPrompt() called with: fep = [org.javarosa.form.api.FormEntryPrompt@5f5c468], context = [org.odk.collect.android.activities.FormEntryActivity@a038933], readOnlyOverride = [false]
I/FormEntryActivity: Created view for group Image widgets Image widget
D/FormEntryActivity: showView() called with: next = [org.odk.collect.android.views.ODKView{3641f8 V.E...... ......I. 0,0-0,0}], from = [FADE]

```

------------------

The collected logs suggests that when `Do not keep activities` is enabled, the activity gets destroyed when switching to camera app for capturing a photo. So, the app gets killed and the widgets gets removed. When the app is reopened after taking the photo, `FormEntryActivity` gets recreated and the first method that gets called after `onCreate()` is `onActivityResult()` and then `onResume()`. This means that `currentView` isn't created by the time the flow reaches to `onActivityResult()` and hence no widget is available for setting the current saved answer to.

I was only able to reproduce this bug only with `Do not keep activities` enabled. So, I believe this bug is only visible on certain devices or with certain camera apps. To fix this, we need to make sure that the widget view is available when the `onActivityResult()` is called. That's why the last commit https://github.com/opendatakit/collect/pull/2538/commits/9fe36ad93c3b6d8e27e8c8f45e59f007db96f43c fixes the bug by recreating the widget just before trying to set the answer.

@grzesiek2010 Does this sound reasonable to you?

grzesiek2010 commented:
@shobhitagarwal1612 

sorry for the late response, holidays you know...

I still don't like the solution it's another `refreshCurrentView()` call so it's rather a workaround as I said.
If we decide that first `refreshCurrentView()` is called in `onResume()` we should perform all operations on the view after it's called. You know that `onActivityResult` is called before so you added your workaround but it's not a good way.

We call `refreshCurrentView()` in many places I would start with investigating if we really need all of them. For example, why we need to call `refreshCurrentView()` at the end of `onActivityResult` when we know it's called again in `onResume()` and `onResume()` is called after `onActivityResult` etc.
understanding all cases and maybe removing some unnecessary lines of code would be better.

shobhitagarwal1612 commented:
@grzesiek2010 That workaround shouldn't be required in the ideal condition. But as caught by you, some of the devices destroy the activity instead of just pausing before starting the image capture intent. 

> We call refreshCurrentView() in many places I would start with investigating if we really need all of them. For example, why we need to call refreshCurrentView() at the end of onActivityResult when we know it's called again in onResume() and onResume() is called after onActivityResult etc.
understanding all cases and maybe removing some unnecessary lines of code would be better.

:+1:

lognaturel commented:
I've finally taken the time to do a deeper dive here because it is an issue when field lists have widgets that call out to other activities. Rebuilding the view multiple times makes it hard to ensure a consistent user experience. I started by making an inventory of when `refreshCurrentView` is called. Here are my notes:
* loadForm if not a new form and formcontroller not null
* end of onActivityResult
  * Introduced in https://github.com/opendatakit/collect/commit/920b6a28dbc35208a0bd232100e3fb2f0946af84 which fixes
    * https://code.google.com/archive/p/opendatakit/issues/619
    * https://code.google.com/archive/p/opendatakit/issues/567
* onResume if there's no ongoing load and an active formController
* repeat dialog shown in field list (is this really what we want?)
* repeat delete
* language change
* loadingComplete -- if there's an activity result, first refresh current view, then call onActivityResult
* loadingComplete -- if we're not allowed to move backwards, jump to index and rebuild the view
* loadingComplete -- end of normal path through the method
* savingComplete -- required answer empty (is this necessary?)
* MediaLoadingTask.onPostExecute

I think overall it's been used as a convenient reset whenever there were view issues and there are a number of uses in that list that seem too aggressive to me. Like you, I focused on the calls in `onActivityResult`, `onResume` and `MediaLoadingTask.onPostExecute`. In #3023, I propose adding some infrastructure for widgets to notify the `FormEntryActivity` of their values being changed. This makes it possible to selectively update parts of the view rather than rebuilding the whole thing. You'll see some similarities with your proposed big refactor although it's not quite as far-reaching.

I'd love to get your thoughts on https://github.com/opendatakit/collect/pull/3023/commits/de88d0cd8d11ecf31aa3a57e6d7f028cfb8e4063 which should address the issue tackled in this PR.

lognaturel commented:
Closing because #3023 addresses the issues here and adds test to verify that related functionality still works.

shobhitagarwal1612 commented:
Sorry, but I didn't get time to review the PR. Thanks for fixing this!
